asteroidsyncserviced/platforms/ubuntutouch/ubuntuplatform.cpp

```diff
@@ -82,18 +95,137 @@ QString UbuntuPlatform::encodeAppName(const QString appName) const
         return "unkown";
 }
 
-void UbuntuPlatform::updateMusicStatus()
+void UbuntuPlatform::onMediaServiceReady()
+{
+    setupMusicService();
+}
+
+void UbuntuPlatform::onPreviousMusicTitle()
+{
+    return sendMusicControlCommand("Previous");
+}
+
+void UbuntuPlatform::onNextMusicTitle()
+{
+    sendMusicControlCommand("Next");
+}
+
+void UbuntuPlatform::onPlayMusicTitle()
+{
+    sendMusicControlCommand("PlayPause");
+}
+
+void UbuntuPlatform::onPauseMusicTitle()
+{
+    sendMusicControlCommand("PlayPause");
+}
+
+void UbuntuPlatform::sendMusicControlCommand(QString method)
+{
+    QDBusMessage call = QDBusMessage::createMethodCall(m_mprisService, "/org/mpris/MediaPlayer2", "org.mpris.MediaPlayer2.Player", method);
+    QDBusError err = QDBusConnection::sessionBus().call(call);
+
+    if (err.isValid()) {
+        qDebug() << "Error calling mpris method on" << m_mprisService << ":" << err.message();
+    }
+}
+
+void UbuntuPlatform::setupMusicService()
+{
+    if (!m_mprisService.isEmpty()) {
+        disconnect(this, SLOT(onMprisPropertiesChanged(QString,QVariantMap,QStringList)));
+    }
+
+    QDBusConnectionInterface *iface = QDBusConnection::sessionBus().interface();
+    const QStringList &services = iface->registeredServiceNames();
+    foreach (QString service, services) {
+        if (service.startsWith("org.mpris.MediaPlayer2.")) {
+            m_mprisService = service;
+            fetchPropertyAsync("Metadata");
+            fetchPropertyAsync("PlaybackStatus");
+            QDBusConnection::sessionBus().connect(m_mprisService, "/org/mpris/MediaPlayer2", "", "PropertiesChanged", this, SLOT(onMprisPropertiesChanged(QString,QVariantMap,QStringList)));
+            break;
+        }
+    }
+}
+
+void UbuntuPlatform::fetchPropertyAsync(const QString &propertyName)
 {
+    if (!m_mprisService.isEmpty()) {
+        QDBusMessage call = QDBusMessage::createMethodCall(m_mprisService, "/org/mpris/MediaPlayer2", "org.freedesktop.DBus.Properties", "Get");
+        call << "org.mpris.MediaPlayer2.Player" << propertyName;
+        QDBusPendingCall pcall = QDBusConnection::sessionBus().asyncCall(call);
+        QDBusPendingCallWatcher *watcher = new QDBusPendingCallWatcher(pcall, this);
+        connect(watcher, &QDBusPendingCallWatcher::finished, [this, propertyName](QDBusPendingCallWatcher *watcher) {
+            watcher->deleteLater();
+            QDBusReply<QDBusVariant> reply = watcher->reply();
+            if (reply.isValid()) {
+                propertyChanged(propertyName, reply.value().variant());
+            } else {
+                qDebug() << "123" << reply.error().message();
```

Comment:
Would you like to use a more expressive debugging message? :)

Reply:
Oh, yes i will change that